### ISSUE TITLE
Problem when trying to click a button that is behind the keyboard

### ISSUE THREAD

**Issue by jamesbluecrow** [SchibstedSpain/Barista#183]
In a test for our login screen, I found an "error" in the library when writing into an `EditText/AutoCompleteTextView`, the "problem" appears when the screen has a button at the bottom and it is hidden behind the keyboard.

![screenshot_1512221877](https://user-images.githubusercontent.com/8219006/33516246-7321ec74-d767-11e7-918f-7dd9e2117cc9.png)
![screenshot_1512221887](https://user-images.githubusercontent.com/8219006/33516247-7334a99a-d767-11e7-9e14-b011ea9f4f6c.png)

We have a test like this:

```
BaristaEditTextInteractions.writeTo(R.id.email, "barista@gmail.com")
BaristaEditTextInteractions.writeTo(R.id.password, "")
BaristaClickInteractions.clickOn(R.id.submit)
```

But we get this is the error because we are trying to click a button that is not visible and it's hidden behind the keyboard.

```
com.schibsted.spain.barista.internal.failurehandler.BaristaException: 
Could not click on view with id: com.trov.maui.staging:id/submit
```

My proposal would be to close by default the keyboard after the text has been written, also it could provide the option to submit from the keyboard.

I've already created some extensions for this so I will open a PR to show the proposed solution.'

**Comment by rocboronat** [comment]
Hey, good issue! By the way, don't you think that you have an issue in your app? I mean, what if you just put the login button below the password, instead of doing it at the bottom of the screen? In that way, you will save a tap to the final customer.

Oh, and send that proposal, for sure! :·D

**Comment by jamesbluecrow** [comment]
hahaha, that is something that we would need to tell to the design team ;) Thanks for the suggestion!

Another option is to support `Submit` from the keyboard (which right now I'm not sure we do).

Here is the PR: https://github.com/SchibstedSpain/Barista/pull/184

**Comment by rocboronat** [comment]
Hey, what do you think about just adding the methods to close the keyboard to the Barista API? In that way, you can close the keyboard, but usually in the same way than you will find a bug in his app thanks to the Espresso test. I mean, your proposal will give a green in a test, but actually what you are doing is bypassing a bug.

Let me think about it after the weekend! :·)

**Comment by jamesbluecrow** [comment]
I'm ok with adding support to close the keyboard as part of the Barista API although I believe that here we are talking about two different things.
The fact that a button is behind the keyboard is not a bug since we can decide how we want the app to resize the screen.
On the other hand I think that I like the fact that the keyboard is close when the action of writing finishes because it was what opened in the first place, although we could do that it doesn't close it by default if you feel it shouldn't.

**Comment by Sloy** [comment]
Hi @jamesbluecrow! In this cae I think your test is helping you see the process your users go though. Barista can't click the button, and your users can't either. They need to close it manually. If that's what you want, I think your test should also go through that step: either with a `pressBack()` or some new `closeKeyboard()`.

I'm trying to think on the opposite use case here. Someone who designs the button to always be visible and clickable, they might expect the test to fail if they change something and the keyboard is blocking the button. Barista does handle scrolling magically, because scroll is such a natural and intuitive thing already. But I'm not so sure about closing the keyboard 🤔

**Comment by jamesbluecrow** [comment]
In fact I don't see it that different from scrolling. In any case I see your point and I understand your concern. What do you think if I modify the code to do this @Sloy @rocboronat ?:
- By default: don't close keyboard (at it was before)
- Add option to close
- Add option to submit

I already added those to options, but I could see that we don't want to execute any of them unless we specify it from the test.

**Comment by Sloy** [comment]
You mean adding some independent Keyboard interactions to close or submit it? 
I think they would be really awesome additions as separate interactions. That way anyone can decide to use how or wherever they want.

I'm imagining a use case like:
```java
writeTo(R.id.edittext, "My text");
keyboardSubmit();
```

What do you think?

**Comment by rocboronat** [comment]
@jamesbluecrow I'm afraid of adding this options as a flags in the method. For me, the best option is to add them as new Keyboard interactions.

IMHO, in your particular case, the test of your activity should be:
```
writeToEditText(R.id.login,"login")
writeToEditText(R.id.password,"password")
closeKeyboard()
clickOn(R.id.button)
```
Because that's what the real user should do. If we automate closing the keyboard, your test will not reflect that issue/feature of your screen: the user has to close the keyboard before tapping the login button.

TL;DR: +1 to create a new class `KeyboardInteractions` with the `closeKeyboard()`,  `clickSubmit()` and `clickNext()` methods. BTW, I'm being creative, I don't know if all the methods I suggest are technically possible 🙌  By the way... I'm sure that `clickSubmit()` method will help apps to improve. Personally, I don't know if some of my apps submit the form when tapping that keyboard button, and adding it to Barista will be the first step to do it 😅

**Comment by rocboronat** [comment]
By the way @jamesbluecrow, sorry for this comment **after** you done the full work 🤕

**Comment by jamesbluecrow** [comment]
I was suggesting to have this:

```
Option 1: writeTo(R.id.login,"login") (don't do anything with the keyboard)
Option 2: writeTo(R.id.login,"login", Action.CLOSE) (close the keyboard)
Option 3: writeTo(R.id.login,"login", Action.SUBMIT) (submit the form)
```

I have a question for you, what is the logic when we do `scrollTo()` to find a view? Maybe closing the keyboard could be done at the same moment instead of having tight to the edittext?
I don't see why we should require in the test one thing but not the other.

If we finally decide to have the devs to write `closeKeyboard()` or `submit()` It should be:

```
KeyboardInteractions.close(R.id.edittext)
KeyboardInteractions.submit(R.id.edittext)
```

Since the `submit()` action is received in the listener for each edittext and in fact it the keyboard is associated to the edittext that is in focus.

Let me know what you think, I have no problem on creating `KeyboardInteractions` instead but to me it seems that we are going to force tests to include steps that could be simplified in a single step and we assume that the keyboard is visible when Android can run in devices that may not have a visible keyboard.

**Comment by rocboronat** [comment]
The `scrollTo()` fixes an issue related to the test environment, not the app itself. The full situation is that you have a green in a test, and you are happy. Then, another day, the test fails just because you are using a smaller device that doesn't show that button. You say **wtf** and blame Espresso 😂 . If you don't have the Barista magic, you should fill **all your tests** with `scrollTo()`. If you have 10 interactions in a test, you must have 10 calls to `scrollTo()`. Lots of interactions that doesn't add value when reading the test... and that's what Barista fixes.

In the other hand, as the `writeTo` interaction doesn't close the keyboard, you found an issue in your app: users are unable to click the login button. So, you have to change your app, or change your test. But not Barista! 😂

**Comment by rocboronat** [comment]
About the need of adding the id of the editttext in the close and submit methods, go on! If possible,  we will try to remove the need of those ids, trying to get the focused view automatically. But let's move on.

**Comment by jamesbluecrow** [comment]
Well, the problem that I'm fixing by closing the keyboard is exactly the same, in my case the button is aligned to the bottom but it could be in the middle of the screen and depending on the screen size it would need to close the keyboard or not.

Anyways, I'll modify the PR to add the interactions for the keyboard :) (and probably keep the extensions locally for our app ;) )

**Comment by rocboronat** [comment]
Thank you @jamesbluecrow ! :·)

**Comment by jamesbluecrow** [comment]
I've created a new PR https://github.com/SchibstedSpain/Barista/pull/185 with a class named `BaristaKeyboardInteractions` that has two functions:
- close()
- pressImeActionButton()

It is possible to close the keyboard without providing a viewId so that's cool, although there is no way to test that the keyboard has been closed without adding some "hacky" code that I'm not sure we want. Let me know what you guys think.

**Comment by jamesbluecrow** [comment]
Some methods to support actions with the keyboard were added.

```
BaristaKeyboardInteractions.closeKeyboard()
BaristaKeyboardInteractions.pressImeActionButton()
```
